### DISCUSSION TITLE
`dbt-core` Python dependency audit

### DISCUSSION THREAD

**Discussion by aranke** [dbt-labs/dbt-core#6915] [Ideas]
A place to discuss dependencies `dbt-core` should include or exclude in the PyPI distribution.
See a current list of included dependencies [here](https://github.com/dbt-labs/dbt-core/blob/main/core/setup.py#L49-L71).

Ground rules:
1. Please create a comment for each type of dependency (e.g., `CLI library`) `dbt` should include or exclude instead of a specific solution (e.g., Click).
2. Please leave questions/comments/concerns about a dependency type or alternatives as a reply to the comment.
3. Developer dependencies (e.g., `pytest` `flake8`, `mypy`, `black`) that are not included in the PyPI distribution are out of scope for this discussion.

Looking forward to a great discussion!

**Comment by aranke** [comment]
# CLI Formatting

We're currently using [`colorama`](https://pypi.org/project/colorama/).

**Reply by aranke** [reply]
I'd like us to move to [Rich](https://pypi.org/project/rich/) for the following reasons:
1. Much more popular (42k ⭐ vs. 3k ⭐)
2. Does more formatting than simply coloring output, including:
    1. Tables
    2. Progress bars and spinners
    3. Syntax highlighting and rich error trace backs
3. Integrates well with our new CLI library [`click`](https://pypi.org/project/click/) via [`rich-click`](https://pypi.org/project/rich-click/).

**Comment by jtcohen6** [comment]
# Serialization & Validation

For certain validation steps, we use [`hologram`](https://github.com/dbt-labs/hologram), a package that we maintain (not very actively). It makes heavy use of [`jsonschema`](https://pypi.org/project/jsonschema/).

For serialization (and some validation), we use a mix of [`mashumaro`](https://pypi.org/project/mashumaro/) and [`betterproto`](https://pypi.org/project/betterproto/) (recent addition).

**Reply by jtcohen6** [reply]
_Proposed changes_

Shorter term:
- https://github.com/dbt-labs/dbt-core/issues/6775

Longer term:
- https://github.com/dbt-labs/dbt-core/issues/6776

**Reply by callum-mcdata** [reply]
Crazy idea: 
- Maybe we consider `pydantic`? It appears that `pydantic` does a combination of both (validation & serialization). The V2 release is currently slated for end of Q1 and contains significant rewrites in Rust to increase performance. But perhaps most interestingly, **it [autoproduces your jsonschema](https://docs.pydantic.dev/usage/schema/) for you**. Gimme Gimme

**Reply by iknox-fa** [reply]
The performance of Pydantic will be a forcing factor here for sure.  We only use mashumaro because it's so much more performant than other options.  IIRC, Pydantic was explored in the past and was found to be orders of magnitude too slow for our use case.  I know none of us like hologram, it's super buggy and doesn't always create valid jsonschema (!).  If we had a solution to fix both of those it would be a big win.

**Reply by iknox-fa** [reply]
It turns out I didn't recall correctly and Pydantic wasn't considered in the earlier work involving parsing performance, however, the [benchmarks done by mashumaro](https://github.com/Fatal1ty/mashumaro#benchmark) suggests that legacy versions would be a significant slowdown. 

[Internal slack discussion](https://dbt-labs.slack.com/archives/C044DK497HS/p1676739258081949) (for reference only, pertinent details are summarized here)

**Reply by aranke** [reply]
It's not as mature as the other options here, but I'll throw [apischema](https://wyfo.github.io/apischema/0.18/) into the mix because:
1. It [claims to be faster](https://wyfo.github.io/apischema/0.18/optimizations_and_benchmark/) than similar libraries like mashumaro.
2. I like the design decision to favor [composition over inheritance](https://wyfo.github.io/apischema/0.18/data_model/#composition-over-inheritance-composed-dataclasses-flattening).

**Reply by aranke** [reply]
Another one that looks promising: https://jcristharif.com/msgspec/

**Comment by jtcohen6** [comment]
# Dates & times

We currently use [`pytz`](https://pypi.org/project/pytz/) for:
- [providing users with timezone logic in the base Jinja context](https://docs.getdbt.com/reference/dbt-jinja-functions/modules#pytz)
- [calculating source freshness](https://github.com/dbt-labs/dbt-core/blob/cbb9117ab997289062cd5bacaf4cfd4466ab9b1e/core/dbt/adapters/base/impl.py#L1056)
- [anonymous usage tracking](https://github.com/dbt-labs/dbt-core/blob/cbb9117ab997289062cd5bacaf4cfd4466ab9b1e/core/dbt/tracking.py#L119)

**Reply by jtcohen6** [reply]
Other projects have moved to using [`zoneinfo`](https://docs.python.org/3/library/zoneinfo.html), which is part of the standard library for Python 3.9+

e.g. https://github.com/python-babel/babel/issues/716

Python 3.8's end of life is 14 October 2024, so we'd need some sort of backward compatibility until then.
